src/providers/hana/qgshanasourceselect.cpp

```diff
@@ -0,0 +1,674 @@
+
+    if ( !values.isEmpty() )
+    {
+      QComboBox *cb = new QComboBox( parent );
+      cb->setItemDelegate( new QStyledItemDelegate( parent ) );
+
+      QStandardItemModel *model = new QStandardItemModel( values.size(), 1, cb );
+
+      int row = 0;
+      for ( const QString &value : values )
+      {
+        QStandardItem *item = new QStandardItem( value );
+        item->setFlags( Qt::ItemIsUserCheckable | Qt::ItemIsEnabled );
+        item->setCheckable( true );
+        item->setData( Qt::Unchecked, Qt::CheckStateRole );
+        model->setItem( row++, 0, item );
+      }
+
+      cb->setModel( model );
+
+      return cb;
+    }
+  }
+
+  if ( index.column() == QgsHanaTableModel::DbtmSrid )
+  {
+    QLineEdit *le = new QLineEdit( parent );
+    le->setValidator( new QIntValidator( -1, 999999, parent ) );
+    return le;
+  }
+
+  return nullptr;
+}
+
+void QgsHanaSourceSelectDelegate::setModelData(
+  QWidget *editor, QAbstractItemModel *model, const QModelIndex &index ) const
+{
+  QComboBox *cb = qobject_cast<QComboBox *>( editor );
+  if ( cb )
+  {
+    if ( index.column() == QgsHanaTableModel::DbtmGeomType )
+    {
+      QgsWkbTypes::Type type = static_cast<QgsWkbTypes::Type>( cb->currentData().toInt() );
+
+      model->setData( index, QgsHanaTableModel::iconForWkbType( type ), Qt::DecorationRole );
+      model->setData( index, type != QgsWkbTypes::Unknown ? QgsWkbTypes::displayString( type ) : tr( "Select…" ) );
+      model->setData( index, type, Qt::UserRole + 2 );
+    }
+    else if ( index.column() == QgsHanaTableModel::DbtmPkCol )
+    {
+      QStandardItemModel *cbm = qobject_cast<QStandardItemModel *>( cb->model() );
+      QStringList cols;
+      for ( int idx = 0; idx < cbm->rowCount(); idx++ )
+      {
+        QStandardItem *item = cbm->item( idx, 0 );
+        if ( item->data( Qt::CheckStateRole ) == Qt::Checked )
+          cols << item->text();
+      }
+
+      model->setData( index, cols.isEmpty() ? tr( "Select…" ) : cols.join( QStringLiteral( ", " ) ) );
+      model->setData( index, cols, Qt::UserRole + 2 );
+    }
+  }
+
+  QLineEdit *le = qobject_cast<QLineEdit *>( editor );
+  if ( le )
+  {
+    QString value( le->text() );
+
+    if ( index.column() == QgsHanaTableModel::DbtmSrid && value.isEmpty() )
+      value = tr( "Enter…" );
+
+    model->setData( index, value );
+  }
+}
+
+void QgsHanaSourceSelectDelegate::setEditorData( QWidget *editor, const QModelIndex &index ) const
+{
+  QString value( index.data( Qt::DisplayRole ).toString() );
+
+  QComboBox *cb = qobject_cast<QComboBox *>( editor );
+  if ( cb )
+  {
+    if ( index.column() == QgsHanaTableModel::DbtmGeomType )
+      cb->setCurrentIndex( cb->findData( index.data( Qt::UserRole + 2 ).toInt() ) );
+
+    if ( index.column() == QgsHanaTableModel::DbtmPkCol &&
+         !index.data( Qt::UserRole + 2 ).toStringList().isEmpty() )
+    {
+      QStringList columns = index.data( Qt::UserRole + 2 ).toStringList();
```

Comment:
```suggestion
      const QStringList columns = index.data( Qt::UserRole + 2 ).toStringList();
```

Reply:
Done in 32b71fc50ba7c0211acb515aec311919140e3e8b.